kvdb-rocksdb/benches/bench_read_perf.rs

```diff
@@ -0,0 +1,213 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Benchmark RocksDB read performance.
+//! The benchmark setup consists in writing `NEEDLES * NEEDLES_TO_HAYSTACK_RATIO` 32-bytes random
+//! keys with random values 150 +/- 30 bytes long. With 10 000 keys and a ratio of 100 we get one
+//! million keys; ideally the db should be deleted for each benchmark run but in practice it has
+//! little impact on the performance numbers for these small database sizes.
+//! Allocations (on the Rust side) are counted and printed.
+//!
+//! Note that this benchmark is not a good way to measure the performance of the database itself;
+//! its purpose is to be a tool to gauge the performance of the glue code, or work as a starting point
+//! for a more elaborate benchmark of a specific workload.
+
+const NEEDLES: usize = 10_000;
+const NEEDLES_TO_HAYSTACK_RATIO: usize = 100;
+
+use std::io;
+use std::time::Instant;
+
+use alloc_counter::{count_alloc, AllocCounterSystem};
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use elastic_array::core_::time::Duration;
+use ethereum_types::H256;
+use rand::{distributions::Uniform, seq::SliceRandom, Rng};
+
+use kvdb_rocksdb::{Database, DatabaseConfig};
+
+#[global_allocator]
+static A: AllocCounterSystem = AllocCounterSystem;
+
+criterion_group!(benches, get, iter);
+criterion_main!(benches);
+
+/// Opens (or creates) a RocksDB database in the `benches/` folder of the crate with one column
+/// family and default options. Needs manual cleanup.
+fn open_db() -> Database {
+	let tempdir_str = "./benches/_rocksdb_bench_get";
+	let cfg = DatabaseConfig::with_columns(Some(1));
+	let db = Database::open(&cfg, tempdir_str).expect("rocksdb works");
+	db
+}
+
+/// Generate `n` random bytes +/- 20%.
+/// The variability in the payload size lets us simulate payload allocation patterns: `DBValue` is
+/// an `ElasticArray128` so sometimes we save on allocations.
+fn n_random_bytes(n: usize) -> Vec<u8> {
+	let mut rng = rand::thread_rng();
+	let variability: i64 = rng.gen_range(0, (n as f64 * 0.2) as i64);
+	let plus_or_minus: i64 = if variability % 2 == 0 { 1 } else { -1 };
+	let range = Uniform::from(0..u8::max_value());
+	rng.sample_iter(&range).take((n as i64 + plus_or_minus * variability) as usize).collect()
+}
+
+/// Writes `NEEDLES * NEEDLES_TO_HAYSTACK_RATIO` keys to the DB. Keys are random, 32 bytes long and
+/// values are random, 120-180 bytes long. Every `NEEDLES_TO_HAYSTACK_RATIO` keys are kept and
```

Comment:
isn't values always 140 bytes long, no?

Reply:
No, they're not. I did some stats collection last spring and the average was slightly different in different parts of the key set (smaller at the start, then with a bunch of weird outliers ~20k). (The choice of `ElasticArray128<u8>` for `DBValue` was a pretty good one I think.)

Reply:
right, I just read the `140` in the call didn't realize `n_bytes` generates +/- 20% of that number.